Specialize getHostName()

2kindsofcs opened the pull request:
Fixes #1006
on windows, getHostName() uses Kernel32Util.getComputerName().
and on Mac, Linux, Solaris, and BSD it returns the value from JNA's LibC class method getHostName().

I have a question. [contributing guide](https://github.com/oshi/oshi/blob/master/CONTRIBUTING.md#update-changelog-again) says "Update CHANGELOG Again" and I'm somewhat confused. Do I have to update changelog at the moment I made a pull request or when my pull request got merged?

dbwiddis commented:
> I have a question. [contributing guide](https://github.com/oshi/oshi/blob/master/CONTRIBUTING.md#update-changelog-again) says "Update CHANGELOG Again" and I'm somewhat confused. Do I have to update changelog at the moment I made a pull request or when my pull request got merged?

The idea here is that the first time you complete the change log you don't know the number of your pull request (you can try to guess by looking at the next sequential issue/pull request number) so you would do another commit to insert the number.  When that document was originally written, the request was to use an "amend" commit to minimize the number of actual commits to the project.  Since then, Github has added a "Squash and merge" option that combines all commits for a PR into one so there's no need to do an "amend" commit (but you're welcome to do so).

dbwiddis commented on oshi-core/src/main/java/oshi/software/os/unix/freebsd/FreeBsdNetworkParams.java:
```diff
@@ -29,6 +29,8 @@
 import org.slf4j.LoggerFactory;
 
+import com.sun.jna.Native;
+import com.sun.jna.platform.unix.LibC;
```
While it ultimately leads here, let's import OSHI's `FreeBsdLibC` class.

dbwiddis commented on oshi-core/src/main/java/oshi/software/os/unix/solaris/SolarisNetworkParams.java:
```diff
@@ -23,11 +23,24 @@
 package oshi.software.os.unix.solaris;
 
+import com.sun.jna.Native;
+import com.sun.jna.platform.unix.LibC;
```
While it ultimately leads here, let's import OSHI's `SolarisLibC` class.

dbwiddis commented on oshi-core/src/main/java/oshi/software/os/mac/MacNetworkParams.java:
```diff
@@ -30,6 +30,8 @@
 import org.slf4j.LoggerFactory;
 
+import com.sun.jna.Native;
+import com.sun.jna.platform.unix.LibC;
```
I was going to suggest the Mac `SystemB` class here, but turns out it doesn't import from `LibC` (like it should) so it seems I need to submit a change to JNA now.  :)  (so this can stay as is for now)

dbwiddis replied:
Actually, this doesn't work if the user doesn't have `gcc` installed.  So change this to `SystemB` and add an `extends LibCAPI` (importing from ` com.sun.jna.platform.unix.LibCAPI`) to OSHI's `SystemB` interface.

dbwiddis commented on oshi-core/src/main/java/oshi/software/os/mac/MacNetworkParams.java:
```diff
@@ -30,6 +30,8 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import com.sun.jna.Native;
```
(Optional) The `// NOSONAR` comment suppresses warnings from SonarQube for using `com.sun` imports (although JNA should be an exception).  If you can move those comments up from 2 lines below (it needs to be on the first import) it will be helpful.

dbwiddis commented on oshi-core/src/main/java/oshi/software/os/linux/LinuxNetworkParams.java:
```diff
@@ -74,6 +76,16 @@ public String getDomainName() {
 
+    @Override
+    public String getHostName() {
+        byte[] hostnameBuffer = new byte[4097];
```
(Since you're eventually extending `LibCAPI` you can use the constant `HOST_NAME_MAX` and add 1 for the null.

dbwiddis commented:
This looks good!   Ready to merge... I might make a few tweaks on the inheritance first, as I realized (belatedly) that we should just make the `CLibrary` class inherit from the JNA `LibCAPI`.  Then there are some functions that we can remove from the `CLibrary`.  Happy to do that unless you want to make yet another change. :)

Did you want to make a changelog entry or are you good to go with this as is?

2kindsofcs commented on oshi-core/src/main/java/oshi/software/os/unix/solaris/SolarisNetworkParams.java:
```diff
@@ -23,7 +23,7 @@
 package oshi.software.os.unix.solaris;
 
-import com.sun.jna.Native;
+import com.sun.jna.Native; // NOSONAR
```
I added a `NOSONAR` footnote.

2kindsofcs commented on oshi-core/src/main/java/oshi/software/os/unix/freebsd/FreeBsdNetworkParams.java:
```diff
@@ -29,7 +29,7 @@
 import org.slf4j.LoggerFactory;
 
-import com.sun.jna.Native; 
+import com.sun.jna.Native; // NOSONAR
```
I added a `NOSONAR` footnote.

dbwiddis commented:
Looks good!  I'll let the CI tests run and tweak the C library inheritance again and merge within the hour.  Thanks for contributing to open source!

2kindsofcs commented:
I checked my code once again and found that 2 files don't have "NOSONAR" footnote while there is `import com.sun.jna.Native`. so I added a footnote for each file. 
and I updated the changelog!  I appreciate your kindness and effort to help someone who wants to contribute to an opensource project but inexperienced. For those a couple of days, I was enjoying adding some code to a project source code that I didn't write and it was truly pleasing and thrilling.
